### ISSUE TITLE
reading value back doesn't work anymore

### ISSUE THREAD

**Issue by wolfpaulus** [KayLerch/alexa-skills-kit-states-java#3]
Reading values back in the User Scope (using a dynamodb) doesn't; work anymore in 1.0.0
But works fine in 0.2.2

**Comment by wolfpaulus** [comment]
This test for instance works in 0.2.2 but fails in 1.0.0

public class UserStateTest {

    static String id = null;

    @Before
    public void initialize() {
        SkillUserDB.enable();
        id = String.valueOf(new Random().nextInt(99999));
    }

    @After
    public void cleanup() {
        SkillUserDB.removeUser(SkillUserDB.fakeSession(id));
    }

    @Test
    public void userTest() throws Exception {
        Session session = SkillUserDB.fakeSession(id);
        assertFalse(SkillUserDB.exists(id));

        SkillUser user = SkillUserDB.create(session);
        assertTrue(id.equals(user.getKey()));
        user.setFirstName("John");
        user.setLastName("Doe");
        user.setLastResponse("How are you today?");
        user.saveState();
        assertTrue(SkillUserDB.exists(id)); // fails with 1.0.0 but works with 0.2.2

        user = SkillUserDB.get(session); // same session.
        assertEquals("John", user.getFirstName());
        assertNotNull(user.getLastResponse());

        // create new session, but use the same user id
        session = SkillUserDB.fakeSession(id);
        user = SkillUserDB.get(session);

        assertEquals("John", user.getFirstName());
        assertEquals(null, user.getLastResponse());

        SkillUserDB.removeUser(session); // same session.
        assertFalse(SkillUserDB.exists(id));
    }
}

**Comment by KayLerch** [comment]
Hi Wolf,

thanks for giving feedback. I'd like to investigate the issue you have but I need a bit more information as it seems you are hiding most of the SDK usage inside the SkillUserDB class.

So, my question is what happens inside SkillUserDB.exists(id)? Is it using the new exists() methods of States SDK 1.0.0 handlers?

**Comment by wolfpaulus** [comment]
Hi Kay,
Thanks so much for looking into this … maybe, I’m just using it in an unintended way.

I have a SkillUser class, which I mainly want to keep in the (Scope = AlexaScope.USER)
SkillUserDB just provides some basic static methods to hide all the DynamoDB stuff.
I have both classes attached.

Moreover, I have the need to create the SkillUser outside of the main Alexa lifecycle, I want to create the user, when he performs account linking. For that I create a fakeSession (again, look at the SkillUserDB class for details.

Happy holidays
Cheers
	Wolf

**Comment by KayLerch** [comment]
Hi Wolf,

did you forget to attach the class files or am I missing something? 
Your use case sounds legit. Also, having above code work for you in 0.2.2. makes me concerned about 1.0.0 as this one should be backward compatible.

**Comment by wolfpaulus** [comment]
I though I had the classes attached,
Let me try again .. if if doesn’t work this took eI will use dropbox.
Let me know

**Comment by KayLerch** [comment]
Sorry, I cannot see any attachments. Try use Gist and post the link to your published script file. Or try do whatever gives me the chance to look at your code ;)

**Comment by wolfpaulus** [comment]
OK try this 
https://www.dropbox.com/s/lm06lxjpx3oue39/state.zip?dl=0 <https://www.dropbox.com/s/lm06lxjpx3oue39/state.zip?dl=0>

**Comment by KayLerch** [comment]
I reviewed your code and I figured out that you work with different userIds in your "fakeSessions". Handlers that save state in USER scope use the sessions userId to create a unique record per user (e.g. an item in dynamo, a file in S3). So writing state with a session having userId 123 results your get-method to return nothing when using another session having userId 234. I saw both in your code, one session with userId = uid and another with userId = token. 

As you are using the accessToken as state object ids anyway, you could also save state in APPLICATION scope because your objects are indexed with an id unique per user. This is how it won't matter which userId is set in your session.

To better understand the consequences of id assignments it's always worth having a look into DynamoDB. You will see what makes up a USER scoped state item vs. an APPLICATION scoped item.

I am still confused why it was working for you in version 0.2.2.

**Comment by wolfpaulus** [comment]
Thanks for you help, but I think I figured that out. The problem is that when I create the user record, I don’t have a real session and therefore also don’t have a user id

Instead I use the token for both token and user id and later when I see the user for the 1st time I rewrite the record, setting the correct user id:


public static Session fakeSession(final String sid, final String uid, final String token) {
    return Session.builder()
            .withSessionId(String.valueOf(sid != null ? sid : "amzn1.echo-api.session." + new Random().nextInt(99999)))
            .withApplication(new Application(PROP.getAppProperty("app_id")))
            .withUser(User.builder()
                    .withUserId(uid != null ? uid : "amzn1.ask.account." + String.valueOf(new Random().nextInt(99999)))
                    .withAccessToken(token != null ? token : String.valueOf(new Random().nextInt(99999)))
                    .build())
            .build();
}

This works with 0.2.2. re-writes the record and also copies it into the session.

public static SkillUser get(final Session session) {
    // try to find the user in the current session
    final String token = session.getUser().getAccessToken();
    try {
        Optional<SkillUser> op = new AlexaSessionStateHandler(session).readModel(SkillUser.class, token);
        if (op != null && op.isPresent()) {
            return op.get();
        }
        if (ENABLED) { // read from DB
            final AmazonDynamoDBClient client = new AmazonDynamoDBClient(CREDENTIALS);
            if (DEV_MODE) {
                client.withEndpoint("http://localhost:8000");
            }
            // try to find the user in the dynamo db
            final AlexaStateHandler handler = new AWSDynamoStateHandler(session, client);
            op = handler.readModel(SkillUser.class, token);
            if (op != null && op.isPresent()) { // copy the user into the session
                final SkillUser su = op.get();
                new AlexaSessionStateHandler(session).writeModel(op.get());
                return SkillUserDB.get(session);
            } else {
                // could be the 1st access and we need to update to the correct uid
                if (!token.equals(session.getUser().getUserId())) {
                    final SkillUser su = SkillUserDB.removeUser(fakeSession(null, token, token));
                    if (su != null) { // at creation time, we assigned uid=token
                        Log.info("New record created for " + token);
                        handler.writeModel(su);
                        Log.info("New record created for " + session.getUser().getAccessToken());
                        return SkillUserDB.get(session);
                    }
                }
            }
        }
    } catch (Exception e) {
        Log.error(e.toString());
    }
    return null;
}

With 1.0.0 a lot of my unit tests fail I started to use the new exists method in 1.0.0 but that also make me wonder why this does not work:

final AmazonDynamoDBClient client = new AmazonDynamoDBClient(CREDENTIALS);
    if (DEV_MODE) {
        client.withEndpoint("http://localhost:8000");
    }
    final AlexaStateHandler handler = new AWSDynamoStateHandler(session, client);
    boolean b = handler.exists(SkillUser.class, token, AlexaScope.USER);
    System.out.println(b);
    Optional<SkillUser> op  = handler.readModel(SkillUser.class,token);
    System.out.println(op.isPresent());
}
true
Could not find state for 'com.techcasita.ask.SkillUser:626981' in session attributes.
false

**Comment by KayLerch** [comment]
There was a bug in my code that should now be fixed. Please upgrade to v1.0.1
